### PR TITLE
feat: Implement Badging API

### LINKED ISSUE #489: [security] Upstream vulnerability in decompress module
**Describe the bug**
There is a NPM high severity warning for kevva/decompress which means `yarn audit` fails on the JS cli - due its being used in the imagemin family of algorithms.

See: https://www.npmjs.com/advisories/1217
Issue here: kevva/decompress#71
Potential mitigation strategy: https://github.com/kevva/decompress/issues/71#issuecomment-594242824

The discussion is somewhat long-winded - with a variety of opinions on the security implications of the matter. For our usage of this library I am not concerned since the library is only used for downloading known binaries.

Nevertheless, the "high" severity warning is disconcerting and I wanted to publish this opinion so that you are aware that we are aware of it. Updates will be found here on this issue as well as at the wg-security channel on our Discord.

### LINKED ISSUE #4489: [feat] Support Badging API
### Describe the problem

Display number of messages on dock or taskbar

### Describe the solution you'd like

https://developer.mozilla.org/en-US/docs/Web/API/Badging_API

windows: 
- https://github.com/microsoft/windows-rs/blob/75400268554bbc1a6c59124ac480bc67ad223835/crates/libs/windows/src/Windows/UI/Notifications/mod.rs#L169
- https://docs.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/badges

macos: https://developer.apple.com/documentation/appkit/nsdocktile
linux unity: https://wiki.ubuntu.com/Unity/LauncherAPI#Count [electron](https://github.com/electron/electron/blob/main/shell/browser/linux/unity_service.cc)

### Alternatives considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ahqsoftwares**
closes #4489

This PR implements badging api for Windows, macOS, Linux, IOS from tao upstream [ref #1002](https://github.com/tauri-apps/tao/pull/1002)

### CODE REVIEW

File: crates/tauri-runtime-wry/src/lib.rs [hunk 1]
```diff
@@ -3096,6 +3127,32 @@ fn handle_user_message<T: UserEvent>(
           WindowMessage::RequestRedraw => {
             window.request_redraw();
           }
+          WindowMessage::SetBadgeCount(_count, _desktop_filename) => {
+            #[cfg(target_os = "ios")]
+            window.set_badge_count(
+              _count.map_or(0, |x| x.clamp(i32::MIN as i64, i32::MAX as i64) as i32),
```
Comment (amrbashir) [inline crates/tauri-runtime-wry/src/lib.rs:3133]: why we clamping here?

File: crates/tauri/src/webview/webview_window.rs [hunk 2]
```diff
@@ -1716,6 +1716,31 @@ impl<R: Runtime> WebviewWindow<R> {
     self.window.start_dragging()
   }
 
+  /// Sets the overlay icon on the taskbar **Windows only**
+  /// Setting the icon to None will remove the overlay icon
+  ///
+  /// The overlay icon can be unique for each window.
+  pub fn set_overlay_icon(&self, icon: Option<Image<'_>>) -> crate::Result<()> {
```
Comment (amrbashir) [inline crates/tauri/src/webview/webview_window.rs:1724]:
```suggestion
  #[cfg(target_os = "windows")]
  #[cfg_attr(docsrs, doc(cfg(target_os = "windows")))]
  pub fn set_overlay_icon(&self, icon: Option<Image<'_>>) -> crate::Result<()> {
```

File: crates/tauri/src/webview/webview_window.rs [hunk 3]
```diff
@@ -1716,6 +1716,31 @@ impl<R: Runtime> WebviewWindow<R> {
...
+  /// - **Windows:** Windows supports arbitrary icon as the badge, use `set_overlay_icon` instead.
+  /// - **iOS:** iOS expects i32, the value will be clamped to i32::MIN, i32::MAX.
+  /// - **Android:** Unsupported.
+  pub fn set_badge_count(&self, count: Option<i64>) -> crate::Result<()> {
+    self.window.set_badge_count(count)
+  }
+
+  /// Sets the taskbar badge label **macOS only**
+  /// Setting the label to `0``will remove the badge
+  pub fn set_badge_label(&self, label: Option<String>) -> crate::Result<()> {
```
Comment (amrbashir) [inline crates/tauri/src/webview/webview_window.rs:1741]:
```suggestion
  #[cfg(target_os = "macos")]
  #[cfg_attr(docsrs, doc(cfg(target_os = "macos")))]
  pub fn set_badge_label(&self, label: Option<String>) -> crate::Result<()> {
```

File: crates/tauri/src/webview/webview_window.rs [hunk 4]
```diff
@@ -1716,6 +1716,31 @@ impl<R: Runtime> WebviewWindow<R> {
...
+  /// The overlay icon can be unique for each window.
+  pub fn set_overlay_icon(&self, icon: Option<Image<'_>>) -> crate::Result<()> {
+    self.window.set_overlay_icon(icon)
+  }
+
+  /// Sets the taskbar badge count
+  /// Setting the count to `0``will remove the badge
+  ///
+  /// ## Platform-specific
+  /// - **Windows:** Windows supports arbitrary icon as the badge, use `set_overlay_icon` instead.
```
Comment (amrbashir) [inline crates/tauri/src/webview/webview_window.rs]:
```suggestion
  /// - **Windows:** Unsupported, use [`WebviewWindow::set_overlay_icon`] instead.
```

File: crates/tauri/src/window/mod.rs [hunk 5]
```diff
@@ -2014,6 +2014,43 @@ tauri::Builder::default()
...
+
+  /// Sets the taskbar badge label **macOS only**
+  /// Setting the label to `0``will remove the badge
+  pub fn set_badge_label(&self, label: Option<String>) -> crate::Result<()> {
+    self
+      .window
+      .dispatcher
+      .set_badge_label(label)
+      .map_err(Into::into)
+  }
```
Comment (amrbashir) [inline crates/tauri/src/window/mod.rs]: same as above comments

File: packages/api/src/window.ts [hunk 6]
```diff
@@ -1605,6 +1605,79 @@ class Window {
     })
   }
 
+  /**
+   * Sets the badge count. It is app wide and not specific to this window.
+   *
+   * #### Platform-specific
+   *
+   * - **Windows**: Use `setOverlayIcon` instead.
```
Comment (amrbashir) [inline packages/api/src/window.ts]:
```suggestion
   * - **Windows**: Unsupported. Use @{linkcode Window.setOverlayIcon} instead.
```

File: packages/cli/index.d.ts [hunk 7]
```diff
@@ -2,6 +2,10 @@
 // SPDX-License-Identifier: Apache-2.0
 // SPDX-License-Identifier: MIT
 
+// Copyright 2019-2024 Tauri Programme within The Commons Conservancy
```
Comment (amrbashir) [inline packages/cli/index.d.ts]: let's revert these changes

File: packages/cli/index.js [hunk 8]
```diff
@@ -2,6 +2,10 @@
 // SPDX-License-Identifier: Apache-2.0
 // SPDX-License-Identifier: MIT
 
+// Copyright 2019-2024 Tauri Programme within The Commons Conservancy
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-License-Identifier: MIT
+
```
Comment (amrbashir) [inline packages/cli/index.js]: same

**Review by amrbashir — CHANGES_REQUESTED** [review]

File: .changes/overlay.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+'@tauri-apps/api': minor:feat
+---
+
+Implement Taskbar Badging APIs for Windows, macOS, IOS, Linux
+
+- `set_badge_count`: Linux, macOS, IOS
+
+## Platform Specific APIs
+- `set_overlay_icon`: Windows
+- `set_badge_label`: macOS
```
Comment (amrbashir) [inline .changes/overlay.md]:
```suggestion
"tauri": minor:feat
"tauri-runtime-wry": minor:feat
"tauri-runtime": minor:feat
'@tauri-apps/api': minor:feat
---

Add badging APIs:

- `Window/WebviewWindow::set_badge_count` for Linux, macOS and IOS.
- `Window/WebviewWindow::set_overlay_icon` for Windows Only.
- `Window/WebviewWindow::set_badge_label`for macOS Only.
```

File: crates/tauri-runtime-wry/src/lib.rs (hunk 1 above)
Comment (ahqsoftwares) [reply crates/tauri-runtime-wry/src/lib.rs:3133]: because in iOS we can only pass i32, but other platforms support i64 ranges

File: packages/cli/index.js (hunk 8 above)
Comment (ahqsoftwares) [reply packages/cli/index.js]: Reverted

File: packages/cli/index.d.ts (hunk 7 above)
Comment (ahqsoftwares) [reply packages/cli/index.d.ts]: Reverted

File: crates/tauri/src/window/mod.rs [hunk 10]
```diff
@@ -2014,6 +2014,47 @@ tauri::Builder::default()
...
+
+  /// Sets the taskbar badge count
+  /// Setting the count to `0``will remove the badge
+  ///
+  /// ## Platform-specific
+  /// - **Windows:** Unsupported, use [`Window::set_overlay_icon`] instead.
+  /// - **iOS:** iOS expects i32, the value will be clamped to i32::MIN, i32::MAX.
+  /// - **Android:** Unsupported.
+  #[cfg(target_os = "macos")]
+  #[cfg_attr(docsrs, doc(cfg(target_os = "macos")))]
```
Comment (amrbashir) [inline crates/tauri/src/window/mod.rs]: this should be on set_badge_label, right?

File: packages/api/src/window.ts [hunk 11]
```diff
@@ -2271,11 +2344,11 @@ function mapMonitor(m: Monitor | null): Monitor | null {
...
-        name: m.name,
-        scaleFactor: m.scaleFactor,
-        position: new PhysicalPosition(m.position),
-        size: new PhysicalSize(m.size)
-      }
+      name: m.name,
+      scaleFactor: m.scaleFactor,
+      position: new PhysicalPosition(m.position),
+      size: new PhysicalSize(m.size)
+    }
```
Comment (amrbashir) [inline packages/api/src/window.ts]: let's revert this formatting

**Review by amrbashir — CHANGES_REQUESTED** [review]

File: packages/api/src/window.ts (hunk 11 above)
Comment (ahqsoftwares) [reply packages/api/src/window.ts]: It was actually done by prettier

File: crates/tauri/src/webview/webview_window.rs [hunk 12]
```diff
@@ -1716,6 +1716,35 @@ impl<R: Runtime> WebviewWindow<R> {
...
+  ///
+  /// The overlay icon can be unique for each window.
+  #[cfg(target_os = "windows")]
+  #[cfg_attr(docsrs, doc(cfg(target_os = "windows")))]
+  pub fn set_overlay_icon(&self, icon: Option<Image<'_>>) -> crate::Result<()> {
+    self.window.set_overlay_icon(icon)
+  }
+
+  /// Sets the taskbar badge count
+  /// Setting the count to `0``will remove the badge
```
Comment (amrbashir) [inline crates/tauri/src/webview/webview_window.rs]:
```suggestion
  /// Sets the taskbar badge count. Setting the count to `0` will remove the badge.
```

File: crates/tauri/src/webview/webview_window.rs [hunk 13]
```diff
@@ -1716,6 +1716,35 @@ impl<R: Runtime> WebviewWindow<R> {
...
+  /// ## Platform-specific
+  /// - **Windows:** Unsupported, use [`WebviewWindow::set_overlay_icon`] instead.
+  /// - **iOS:** iOS expects i32, the value will be clamped to i32::MIN, i32::MAX.
+  /// - **Android:** Unsupported.
+  pub fn set_badge_count(&self, count: Option<i64>) -> crate::Result<()> {
+    self.window.set_badge_count(count)
+  }
+
+  /// Sets the taskbar badge label **macOS only**
+  /// Setting the label to `0``will remove the badge
```
Comment (amrbashir) [inline crates/tauri/src/webview/webview_window.rs]: I think you meant
```suggestion
  /// Sets the taskbar badge label **macOS only**. Using `None` will remove the badge.
```

File: crates/tauri/src/window/mod.rs [hunk 14]
```diff
@@ -2014,6 +2014,47 @@ tauri::Builder::default()
...
+  pub fn set_overlay_icon(&self, icon: Option<Image<'_>>) -> crate::Result<()> {
+    self
+      .window
+      .dispatcher
+      .set_overlay_icon(icon.map(|x| x.into()))
+      .map_err(Into::into)
+  }
+
+  /// Sets the taskbar badge count
+  /// Setting the count to `0``will remove the badge
```
Comment (amrbashir) [inline crates/tauri/src/window/mod.rs]:
```suggestion
  /// Sets the taskbar badge count. Setting the count to `0` will remove the badge.
```

File: crates/tauri/src/window/mod.rs [hunk 15]
```diff
@@ -2014,6 +2014,47 @@ tauri::Builder::default()
...
+  pub fn set_badge_count(&self, count: Option<i64>) -> crate::Result<()> {
+    self
+      .window
+      .dispatcher
+      .set_badge_count(count, Some(format!("{}.desktop", self.package_info().name)))
+      .map_err(Into::into)
+  }
+
+  /// Sets the taskbar badge label **macOS only**
+  /// Setting the label to `0``will remove the badge
```
Comment (amrbashir) [inline crates/tauri/src/window/mod.rs]:
```suggestion
  /// Sets the taskbar badge label **macOS only**. Using `None` will remove the badge.
```

File: packages/api/src/window.ts [hunk 16]
```diff
@@ -1605,6 +1605,79 @@ class Window {
...
+   *
+   * - **Windows**: Unsupported. Use @{linkcode Window.setOverlayIcon} instead.
+   *
+   * @example
+   * ```typescript
+   * import { getCurrentWindow } from '@tauri-apps/api/window';
+   * await getCurrentWindow().setBadgeCount(5);
+   * ```
+   *
+   * @param count The badge count. Set undefined to remove the badge.
```
Comment (amrbashir) [inline packages/api/src/window.ts]:
```suggestion
   * @param count The badge count. Use `undefined` to remove the badge.
```

File: packages/api/src/window.ts [hunk 17]
```diff
@@ -1605,6 +1605,79 @@ class Window {
...
+  /**
+   * Sets the badge cont **macOS only**.
+   *
+   * @example
+   * ```typescript
+   * import { getCurrentWindow } from '@tauri-apps/api/window';
+   * await getCurrentWindow().setBadgeLabel("Hello");
+   * ```
+   *
+   * @param label The badge label. Set undefined to remove the badge.
```
Comment (amrbashir) [inline packages/api/src/window.ts]:
```suggestion
   * @param label The badge label. Use `undefined` to remove the badge.
```

**Review by amrbashir — CHANGES_REQUESTED** [review]
This is ready to merge, just a couple of docs nit-picks and you need to regen the global script, just run `pnpm build:api` in root

File: packages/api/src/window.ts (hunk 11 above)
Comment (amrbashir) [reply packages/api/src/window.ts]: looks like something is wrong on your setup, as CI was failing because of this

**Review by amrbashir — APPROVED** [review]

**Comment by amrbashir** [comment]
Thank you
